### ISSUE TITLE
Trait bounds incorrectly reported, causing error in const-fn setting

### ISSUE THREAD

**Issue by philliard3** [rust-lang/rust#88377]
I tried this code:

```rust
#![feature(const_fn_trait_bound)]

use std::marker::PhantomData;

pub trait KeySchema {
    const KEYS: &'static [usize];
}

const fn successor_len<KS: KeySchema>() -> usize {
    KS::KEYS.len() + 1
}

pub struct KeyCons<Tail> {
    _tail: PhantomData<Tail>,
}

impl<Tail> KeyCons<Tail>
where
    Tail: KeySchema,
{
    // Works if removing `const`:
    // pub fn succ_len() -> usize {
    pub const fn succ_len() -> usize {
        successor_len::<Tail>()
        // or it will work by instead inlining the contents of successor_len::<Tail>()
        // Tail::KEYS.len() + 1
    }
}
```
[( Playground )](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=0eb10255633ed5ba4461bcc08ed363b4)

I expected to see this happen:
I expected that this would behave as though this were a non-const function, and successor_len::<Tail>() would cause the compiler to inline the  implementation of successor_len.

Instead, this happened: 
The compiler claims that the KeySchema bound doesn't exist on Tail, and it produces an error instead of executing the call and inlining the const fn as expected.

```rust
error[E0277]: the trait bound `Tail: KeySchema` is not satisfied
  --> src/lib.rs:24:25
   |
24 |         successor_len::<Tail>()
   |                         ^^^^ the trait `KeySchema` is not implemented for `Tail`
   |
note: required by a bound in `successor_len`
  --> src/lib.rs:9:28
   |
9  | const fn successor_len<KS: KeySchema>() -> usize {
   |                            ^^^^^^^^^ required by this bound in `successor_len`
help: consider further restricting this bound
   |
19 |     Tail: KeySchema + KeySchema,
   |                     +++++++++++

For more information about this error, try `rustc --explain E0277`.
error: could not compile `playground` due to previous error
```

### Meta

This failure to compile currently happens on the playground
version 1.56.0-nightly (2021-08-25 0afc20860eb98a29d9bb)
This is on Playground

This code does compile on my local version
rustc 1.56.0-nightly (25b764849 2021-08-04)
This is on Windows

Discussion on The Rust Programming Language Discord server leads me to believe this is a bug and a regression from a previous nightly version.

**Comment by philliard3** [comment]
This appears to be fixed as of nightly version: 1.56.0-nightly (2021-09-01 50171c310cd15e1b2d37)

As a result, I am closing this issue.
